### PR TITLE
Chapter 5: MicroProfile Configuration

### PR DESCRIPTION
**Pull request by ttelang**
Adding Chapter 5 about MicroProfile Configuration

### CODE REVIEW

**Comment by Emily-Jiang** [comment]
I am reviewing this PR and will provide feedback asap.

**Comment by ttelang** [comment]
Awaiting review feedback and approval.

File: chapter05/chapter05.adoc [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+
+The MicroProfile Configuration specifications offer a set of APIs that
+enable you to handle your application’s configuration efficiently. They
+allow you to easily manage and customize your application’s
+configurations, making it a valuable tool for developers.
+
+The MicroProfile Configuration APIs provide the following capabilities
+for managing the configuration settings of your application:
+
+- It allows reading and writing configuration values.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:112]:
```suggestion
- It allows reading configuration values.
```

File: chapter05/chapter05.adoc [hunk 2]
```diff
@@ -0,0 +1,596 @@
...
+- It allows applications to retrieve configuration values reliably,
+supporting various sources, such as property files, system properties,
+environment variables, and more.
+- Observe changes to configuration values using config events, making
+tracking and managing changes to your application’s configuration
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:117]: The line 116-118 was incorrect. please delete.

File: chapter05/chapter05.adoc [hunk 3]
```diff
@@ -0,0 +1,596 @@
...
+allows retrieving the static instance of the Config object.
+- `ConfigBuilder` - An interface used to create a Config instance manually.
+It can add default sources, converters, and configuration sources.
+- `ConfigSource` - This class represents a source of configuration values.
+It reads configuration data from a specific source, such as system
+properties, environment variables, files, or data stores.
+- `Property` - It represents a key/value pair in the configuration data.
+- `Converter<T>` - This interface implements custom converters that convert
+configuration values from String to any desired type.
+- `PropertyExpression` - An interface supporting expression evaluation
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:137]: This is incorrect. There is no such an interface.

File: chapter05/chapter05.adoc [hunk 4]
```diff
@@ -0,0 +1,596 @@
...
+`ConfigProperty` also supports type conversion, so we can inject our configuration data into fields of any type:
+
+[source, java]
+----
+   @Inject
+   @ConfigProperty(name="product.maintenanceMode", defaultValue="false")
+   private boolean maintenanceMode;
+----
+
+In this example, the `product.maintenanceMode` property will be converted to an `Boolean` before it is injected into the `maintenanceMode` field. If the conversion fails, a `org.eclipse.microprofile.config.ConversionFailedException` will be thrown.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:307]: The exception does not exist though.

File: chapter05/chapter05.adoc [hunk 5]
```diff
@@ -0,0 +1,596 @@
...
+}
+
+
+public class MaintenanceMessage {
+    @ConfigProperty(name="product.maintenanceMessage")
+    private String message;
+}
+----
+
+In this example, we’re injecting a property named "product.maintenanceMessage" into the message field of our MaintenanceMessage class. If the property does not exist, a `org.eclipse.microprofile.config.ConfigPropertyNotFoundException` will be thrown.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:330]: The exception does not exist.

File: chapter05/chapter05.adoc [hunk 6]
```diff
@@ -0,0 +1,596 @@
...
+- `Map<String, String> getProperties()` : Returns a map of the properties in this ConfigSource. The keys in the map are the property names, and the values are the property values. 
+- `getValue(String propertyName)` : Returns the value of the given property. If the property is not found, this method returns null. 
+- `Set getPropertyNames()` : Returns a Set of the property names in this ConfigSource.
+
+Let’s implement a feature in our MicroProfile e-Commerce application to integrate payment gateway configuration dynamically by creating a PaymentServiceConfigSource (a custom ConfigSource) which could fetch API keys and endpoints. This would ensure that payment service configurations are up-to-date and can be changed without redeploying the application.
+
+MicroProfile Config provides a flexible framework for integrating with various external configuration providers, enabling applications to load configurations from sources beyond the default system properties, environment variables, and microprofile-config.properties files. This capability is crucial for modern applications that may need to pull configuration from dynamic sources like cloud services, databases, or custom APIs.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:345]: This paragraph is out of place. It should be placed in the beginning of this chapter.

File: chapter05/chapter05.adoc [hunk 7]
```diff
@@ -0,0 +1,596 @@
...
+
+=== Enabling MicroProfile Config in Open Liberty
+
+Open Liberty requires a `server.xml` file for server configuration. This file should be located at `/src/main/liberty/config/server.xml` within your project. To enable MicroProfile Config, you need to include the `mpConfig` feature in the `<featureManager>` section.
+
+[source, xml]
+----
+<server description="MicroProfile Tutorial Server">
+  <featureManager>
+    <feature>mpConfig-1.4</feature> <!-- Use the appropriate version -->
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:499]:
```suggestion
    <feature>mpConfig-3.1</feature> <!-- Use the appropriate version -->
```

File: chapter05/chapter05.adoc [hunk 8]
```diff
@@ -0,0 +1,596 @@
...
+
+=== Strategies for Handling Configuration Change Events
+
+Although direct support for configuration change events is not provided by MicroProfile Config specification itself, applications can implement their mechanisms or use external libraries to achieve this functionality. To implement dynamic updates in your MicroProfile Config usage, you might need to adopt one of the following approaches:
+
+- *Manual Refresh:* Provide a mechanism (e.g., an admin-restricted endpoint) to manually trigger a refresh of the configuration. This approach gives control over when changes are applied but requires manual intervention.
+- *Polling*: Implement a scheduler that periodically checks certain configuration properties for changes. This approach is straightforward but might introduce latency between the actual change and its detection.
+- *Event-driven Updates*: If your configuration source supports event notifications (for example, a database trigger or a cloud service event), you can set up listeners that update your application’s configuration in response to these events.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:546]: Not sure how you can do this. There is no api for such notification.

File: chapter05/chapter05.adoc [hunk 9]
```diff
@@ -0,0 +1,596 @@
...
+
+While the MicroProfile Config specification provides a powerful and flexible framework for configuration management, handling dynamic updates and configuration change events may require additional custom development or the use of external tools. By considering the strategies mentioned above, developers can effectively manage configuration changes, ensuring their microservices remain responsive and resilient in dynamic environments.
+
+While MicroProfile Config provides a flexible and powerful framework for managing application configurations, securing sensitive data within those configurations requires careful consideration and the implementation of additional security measures. By following best practices for configuration security, developers can help protect their applications from potential threats and vulnerabilities.
```
Comment (Emily-Jiang) [inline chapter05/chapter05.adoc:589]: This paragraph is quite repetitive and repeat the beginning of the message in line 587.

File: chapter05/chapter05.adoc (hunk 6 above)
Comment (ttelang) [reply chapter05/chapter05.adoc]: Fixed by 936f1ce31973eceb2c7eacf12deadb5ea3943b35 and 18da76cb09e9f2807fc28603c48d31b78bd468cf. The idea is to help readers understand and appreciate this capability with the help of the above code example.

**Comment by Emily-Jiang** [comment]
@ttelang please comment here after you have addressed all of the comments

File: chapter05/chapter05.adoc (hunk 5 above)
Comment (ttelang) [reply chapter05/chapter05.adoc:325]: Removed.

File: chapter05/chapter05.adoc (hunk 9 above)
Comment (ttelang) [reply chapter05/chapter05.adoc]: Shortened to be more precise and remove repetition.

**Comment by ttelang** [comment]
I have addressed all of the comments.

File: chapter05/chapter05.adoc (hunk 4 above)
Comment (Emily-Jiang) [reply chapter05/chapter05.adoc:302]: This was not resolved.

File: chapter05/chapter05.adoc (hunk 5 above)
Comment (Emily-Jiang) [reply chapter05/chapter05.adoc:325]: I still see this exception on the line 325.

**Comment by Emily-Jiang** [comment]
> I have addressed all of the comments. @Emily-Jiang - Kindly merge this PR

I double checked and found 2 comments were not resolved. Please address them. Thanks!

**Pull request merged**